packages/firestore/src/api/database.ts

```diff
@@ -1177,150 +1186,179 @@ export class DocumentReference<T = firestore.DocumentData>
+        .getDocumentFromLocalCache(this._key)
+        .then(
+          doc =>
             new DocumentSnapshot(
               this.firestore,
               this._key,
               doc,
-              snapshot.fromCache,
-              snapshot.hasPendingWrites,
+              /*fromCache=*/ true,
+              doc instanceof Document ? doc.hasLocalMutations : false,
               this._converter
             )
-          );
-        }
-      },
-      error: errHandler
-    });
-    const internalListener = this._firestoreClient.listen(
-      InternalQuery.atPath(this._key.path),
-      asyncObserver,
-      options
-    );
-
-    return () => {
-      asyncObserver.mute();
-      this._firestoreClient.unlisten(internalListener);
-    };
+        );
+    } else {
+      return getDocViaSnapshotListener(
+        this._firestoreClient,
+        this,
+        options
+      ).then(snapshot => this._convertToDocSnapshot(snapshot));
+    }
   }
 
-  get(options?: firestore.GetOptions): Promise<firestore.DocumentSnapshot<T>> {
-    validateBetweenNumberOfArgs('DocumentReference.get', arguments, 0, 1);
-    validateGetOptions('DocumentReference.get', options);
-    return new Promise(
-      (resolve: Resolver<firestore.DocumentSnapshot<T>>, reject: Rejecter) => {
-        if (options && options.source === 'cache') {
-          this.firestore
-            .ensureClientConfigured()
-            .getDocumentFromLocalCache(this._key)
-            .then(doc => {
-              resolve(
-                new DocumentSnapshot(
-                  this.firestore,
-                  this._key,
-                  doc,
-                  /*fromCache=*/ true,
-                  doc instanceof Document ? doc.hasLocalMutations : false,
-                  this._converter
-                )
-              );
-            }, reject);
-        } else {
-          this.getViaSnapshotListener(resolve, reject, options);
-        }
-      }
-    );
+  withConverter<U>(
+    converter: firestore.FirestoreDataConverter<U>
+  ): firestore.DocumentReference<U> {
+    return new DocumentReference<U>(this._key, this.firestore, converter);
   }
 
-  private getViaSnapshotListener(
-    resolve: Resolver<firestore.DocumentSnapshot<T>>,
-    reject: Rejecter,
-    options?: firestore.GetOptions
-  ): void {
-    const unlisten = this.onSnapshotInternal(
-      {
-        includeMetadataChanges: true,
-        waitForSyncWhenOnline: true
-      },
-      {
-        next: (snap: firestore.DocumentSnapshot<T>) => {
-          // Remove query first before passing event to user to avoid
-          // user actions affecting the now stale query.
-          unlisten();
+  /**
+   * Converts a ViewSnapshot that contains the current document to a
+   * DocumentSnapshot.
+   */
+  private _convertToDocSnapshot(snapshot: ViewSnapshot): DocumentSnapshot<T> {
+    debugAssert(
+      snapshot.docs.size <= 1,
+      'Too many documents returned on a document query'
```

Comment:
This message is odd relative to the check being done. Can you add a comment with some context?

Reply:
Changed to `Expected zero or a single result on a document-only query`.